Introducing Object Cores (#59075)
zephyrproject-rtos/zephyr: doc/kernel/object_cores/index.rst

.. _object_cores_api:

Object Cores
############

Object cores are a kernel debugging tool that can be used to both identify and
perform operations on registered objects.

.. contents::
    :local:
    :depth: 2

Object Core Concepts
********************

> Each instance of an object has embedded within it an object core field named

Reviewer: nit:

```suggestion
Each instance of an object embeds an object core field named
```

Author: Fixing.

`obj_core`. Objects of the same type are linked together via their respective
object cores to form a singly linked list. Each object core also links to the
their respective object type. Each object type contains a singly linked list
linking together all the object cores of that type. Object types are also
linked together via a singly linked list. Together, this can allow debugging
tools to traverse all the objects in the system.

> Object cores have been integrated into following kernel objects:
>  * Condition Variables
>  * Events
>  * FIFOs and LIFOs
>  * Mailboxes
>  * Memory Slabs
>  * Message Queues
>  * Mutexes
>  * Pipes
>  * Semaphores
>  * Timers
>  * System Memory Blocks

Reviewer: the leading white spaces is causing the list to render as a definition list, so you need to drop it. Also would be nice to link to doc pages (I've added explicit aliases for the pages to keep the list easy to read when browsing the source.

```suggestion
Object cores have been integrated into following kernel objects:

* :ref:`Condition Variables <condvar>`
* :ref:`Events <events>`
* :ref:`FIFOs <fifos_v2>` and ref:`LIFOs <lifos_v2>`
* :ref:`Mailboxes <mailboxes_v2>`
* :ref:`Memory Slabs <memory_slabs_v2>`
* :ref:`Message Queues <message_queues_v2>`
* :ref:`Mutexes <mutexes_v2>`
* :ref:`Pipes <pipes_v2>`
* :ref:`Semaphores <semaphores_v2>`
* :ref:`Timers <timers_v2>`
* :ref:`System Memory Blocks <sys_mem_blocks>`
```

Author: Fixing.

Reviewer: > Fixing.

thx! If you have a chance please test locally: I might have gotten some references wrong as i provided my feedback through GitHub UI and did not actually test this change

Developers are free to integrate them if desired into other objects within
their projects.

Object Core Statistics Concepts
*******************************
A variety of kernel objects allow for the gathering and reporting of statistics.
Object cores provide a uniform means to retrieve that information via object
core statistics. When enabled, the object type contains a pointer to a
statistics descriptor that defines the various operations that have been
enabled for interfacing with the object's statistics. Additionally, the object
core contains a pointer to the "raw" statistical information associated with
that object. Raw data is the raw, unmanipulated data associated with the
statistics. Queried data may be "raw", but it may also have been manipulated in
some way by calculation (such as determining an average).

> Object core statistics have been integrated into the following kernel objects:

Reviewer: typo / copy-paste error?

```suggestion
```

Author: Fixing.

An object type is defined using a global variable of type
:c:struct:`k_obj_type`. It must be initialized before any objects of that type
are initialized. The following code shows how a new object type can be
initialized for use with object cores and object core statistics.

.. code-block:: c

    /* Unique object type ID */

    #define K_OBJ_TYPE_MY_NEW_TYPE  K_OBJ_TYPE_ID_GEN("UNIQ")
    struct k_obj_type  my_obj_type;

    struct my_obj_type_raw_info {
        ...
    };

    struct my_obj_type_query_stats {
        ...
    };

    struct my_new_obj {
        ...
        struct k_obj_core obj_core;
        struct my_obj_type_raw_info  info;
    };

    struct k_obj_core_stats_desc my_obj_type_stats_desc = {
        .raw_size = sizeof(struct my_obj_type_raw_stats),
        .query_size = sizeof(struct my_obj_type_query_stats),
        .raw = my_obj_type_stats_raw,
        .query = my_obj_type_stats_query,
        .reset = my_obj_type_stats_reset,
        .disable = NULL,    /* Stats gathering is always on */
        .enable = NULL,     /* Stats gathering is always on */
    };

    void my_obj_type_init(void)
    {
        z_obj_type_init(&my_obj_type, K_OBJ_TYPE_MY_NEW_TYPE,

>                         offsetof(struct my_new_type, obj_core);

Reviewer 2: Shouldn't this be `struct my_new_obj` instead of `struct my_new_type`? I.e. what you have on line 93.

Reviewer: @peter-mitsis would it make things easier to add an actual code sample demonstrating this? This would help keep the code free of typos, and you wwould still be embed to embed snippets of it in this documentation page using `.. literalinclude:` if you wanted to.

Author: @jhedberg ... you are correct. I will be fixing this shortly.

@kartben ... I am working on a sample project for this.
